### PR TITLE
cloud provider: fix the fake cloud provider

### PR DESCRIPTION
**Pull request by nicolehanjing**
fix the fake cloud provider and update the unit tests for cloud node controller

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fix the fake cloud provider and update the unit tests for cloud node controller
After this merge, back port it into v1.19

To implement instancesV2 interface (https://github.com/kubernetes/cloud-provider-aws/pull/131), we need to handle empty InstancesV2 providerID from getProviderID. A bug was found and fixed: https://github.com/kubernetes/kubernetes/pull/95342

The test case we're now missing in our unit tests for cloud node controller is when existingNode does not have `spec.providerID` already set, the updated node should set providerID based on the providerID returned from `InstanceMetadata`. 
Testing that would require the fake cloud provider to return a generated provider ID not derived from what's already set on the node

After this fix, the fake cloud provider will return ProviderID in InstanceMetadata from the node spec if it exists, otherwise it returns a generated providerID like fake://<ext-id> where ext-id is fetched from the ExtID map

**Which issue(s) this PR fixes**:

Part of #125

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 1]
```diff
@@ -321,8 +321,15 @@ func (f *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprov
 	f.addCall("instance-metadata-by-provider-id")
 	f.addressesMux.Lock()
 	defer f.addressesMux.Unlock()
+
+	providerID := node.Spec.ProviderID
+	_, ok := f.ExtIDErr[types.NodeName(node.Name)]
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:326]: This should probably check `ExtID` instead of `ExtIDErr`, and fake cloud provider could likely use a new field like `MetadataErr` to mock errors when calling `InstanceMetadata`

**Review by andrewsykim — COMMENTED** [review]
/ok-to-test

File: staging/src/k8s.io/cloud-provider/fake/fake.go (hunk 1 above)
Comment (nicolehanjing) [reply staging/src/k8s.io/cloud-provider/fake/fake.go:326]: I thought if there is `ExtIDErr` associated with the nodeName, that means `ExtID` is an empty map, but if `ExtIDErr` doesn't exist, we have ExtID available.
But you are right, maybe checking `ExtID` is more straightforward
will update!

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 2]
```diff
@@ -321,13 +322,20 @@ func (f *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprov
 	f.addCall("instance-metadata-by-provider-id")
 	f.addressesMux.Lock()
 	defer f.addressesMux.Unlock()
+
+	providerID := node.Spec.ProviderID
+	id, ok := f.ExtID[types.NodeName(node.Name)]
+	if providerID == "" && ok {
+		providerID = "fake://" + id
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:329]: On second thought, I think we want `Cloud` to have a field `ProviderID` that holds the value that would be returned here. This way you can define the fake provider ID from the test instead of it being opaque. What do you think?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go [hunk 3]
```diff
@@ -702,6 +702,7 @@ func Test_syncNode(t *testing.T) {
 							Effect: v1.TaintEffectNoSchedule,
 						},
 					},
+					ProviderID: "fake://12345",
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:705]: For reference, this was originally removed here https://github.com/kubernetes/kubernetes/pull/95342/files#r501940685 since the fake cloud provider did not support pre-existing provider IDs

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go [hunk 4]
```diff
@@ -436,14 +436,14 @@ func Test_syncNode(t *testing.T) {
 					CreationTimestamp: metav1.Date(2012, 1, 1, 0, 0, 0, 0, time.UTC),
 				},
 				Spec: v1.NodeSpec{
-					ProviderID: "fake://12345",
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:439]: This change here seems not related, can we undo this just to keep the git history clean?

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:439]: (move line 446 back here)

Comment (nicolehanjing) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:439]: ah sure, good catch!

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 5]
```diff
@@ -321,13 +323,20 @@ func (f *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprov
 	f.addCall("instance-metadata-by-provider-id")
 	f.addressesMux.Lock()
 	defer f.addressesMux.Unlock()
+
+	providerID := node.Spec.ProviderID
+	_, ok := f.ExtID[types.NodeName(node.Name)]
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:328]: Realiziing now that similar to ExtID, `ProviderID` should probably be of type `map[types.NodeName]string`. And here we should check `f.ProviderID` instead of `f.ExtdID`.

Comment (nicolehanjing) [reply staging/src/k8s.io/cloud-provider/fake/fake.go]: gotcha! Will update

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 6]
```diff
@@ -321,13 +323,20 @@ func (f *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprov
 	f.addCall("instance-metadata-by-provider-id")
 	f.addressesMux.Lock()
 	defer f.addressesMux.Unlock()
+
+	providerID := node.Spec.ProviderID
+	id, ok := f.ProviderID[types.NodeName(node.Name)]
+	if providerID == "" && ok {
+		providerID = id
```
Comment (nicolehanjing) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:330]: @andrewsykim updated! PTAL :)

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go (hunk 3 above)
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:705]: This test case name is "provider ID not implemented", so I think the existing node is not supposed to have a provider ID and the updated node is not supposed to have it either.

**Review by andrewsykim — COMMENTED** [review]
One more minor comment, lgtm otherwise

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 7]
```diff
@@ -321,13 +323,20 @@ func (f *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprov
 	f.addCall("instance-metadata-by-provider-id")
 	f.addressesMux.Lock()
 	defer f.addressesMux.Unlock()
+
+	providerID := node.Spec.ProviderID
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:327]: I don't think the fake cloud provider should inherit what's in `node.Spec.ProviderID`. That is actually logic in the cloud node controller that we want to test against.

Comment (nicolehanjing) [reply staging/src/k8s.io/cloud-provider/fake/fake.go:327]: ahh gotcha, updated!

File: staging/src/k8s.io/cloud-provider/fake/fake.go [hunk 8]
```diff
@@ -321,13 +323,20 @@ func (f *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprov
 	f.addCall("instance-metadata-by-provider-id")
 	f.addressesMux.Lock()
 	defer f.addressesMux.Unlock()
+
+	providerID := ""
```
Comment (nicolehanjing) [inline staging/src/k8s.io/cloud-provider/fake/fake.go:327]: @andrewsykim so I make this default empty providerID instead of inheriting from `node.spec.providerID`
PTAL :)

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm
